### ISSUE TITLE
Add the ability to correct observer information

### ISSUE THREAD

**Issue by ehsteve** [sunpy/sunpy#3110]
### Description

It is now abundantly clear that many of the data sources that we support do not provide accurate observer information if it is provided at all (see #3057, #3038, #3037).

Thankfully JPL Horizons can provide satellite location information for many (if not all) data sources supported by SunPy. Also, [astroquery](https://astroquery.readthedocs.io/en/latest/jplhorizons/jplhorizons.html) provides a Python interface to that information. There is the limitation that it may not provide information for recent positions (~1 month). I would suggest that the map factory try to correct or add observer information on when the map is instantiated.

### Additional context

Correct observer information is necessary for the coordinate functionality to work accurately.

**Comment by ayshih** [comment]
Query for SOHO's location via `astroquery` would look something like this:
```python
>>> from astropy.time import Time
>>> from astroquery.jplhorizons import Horizons
>>> from astropy.coordinates import SkyCoord, CartesianRepresentation

>>> obstime = Time('2012-06-01 11:22:33')
>>> soho = Horizons(id='SOHO', id_type='majorbody', location='500@0', epochs=obstime.jd)
>>> soho_vector = CartesianRepresentation(soho.vectors()[0]['x', 'y', 'z'])*u.AU
>>> soho_coord = SkyCoord(CartesianRepresentation(soho_vector)*u.AU, frame='icrs', obstime=obstime)
```

**Comment by Cadair** [comment]
> I would suggest that the map factory try to correct or add observer information when the map is instantiated if it is known that the data is missing or likely incorrect.

No.

We could provide a utility for doing this, although I am not in favour of adding an astroquery dep to core. So maybe just an example?

**Comment by dstansby** [comment]
Here's a working example:
```python
import astropy.io

import sunpy.map
from sunpy.coordinates import get_horizons_coord
from sunpy.map.header_helper import get_observer_meta
from sunpy.net import Fido
from sunpy.net import attrs as a

# Download a single LASCO file
result = Fido.search(a.Instrument('lasco'), a.Time('2010-01-01 00:00:00', '2010-01-01 00:10:00'))

# Load, and print the assumed observer coordinate
filepath = Fido.fetch(result[0, 0])[0]
lasco_map = sunpy.map.Map(filepath)
print(lasco_map.observer_coordinate)

# Get actual coordinate using JPL Horizons
soho_coord = get_horizons_coord('SOHO', time=lasco_map.date)
print(soho_coord)

# Re-load, correcting observer metadata this time
observer_meta = get_observer_meta(soho_coord)
with astropy.io.fits.open(filepath) as hdul:
    data, header = hdul[0].data, hdul[0].header

header.update(observer_meta)
lasco_map = sunpy.map.Map(data, header)
print(lasco_map.observer_coordinate)
```

**Comment by ayshih** [comment]
As I said elsewhere (https://github.com/sunpy/sunpy/issues/3037#issuecomment-1509826224), I think this broad issue should be closed.  Any known problems with a source's observer information should be handled by the source-specific subclass, and thus the need to use `astroquery` or anything else would be a source-specific decision.  The Map factory shouldn't do anything.

**Comment by dstansby** [comment]
I agree that the map factor shouldn't do this. I'm still +0.5 that a utility function that a user can call to set the observer information from JPL Horizons would be within scope of sunpy core though.

**Comment by ayshih** [comment]
Sure, and that's covered by #4189

**Comment by nabobalis** [comment]
Then should this issue be retitled?

**Comment by wtbarnes** [comment]
I've retitled this to be a bit more specific to the discussion that has followed.
